apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/CachingHttpInterceptor.kt

```diff
@@ -48,17 +48,29 @@ class CachingHttpInterceptor(
         return networkMightThrow(request, chain, cacheKey)
       }
       NETWORK_FIRST -> {
+        var networkException: ApolloException? = null
         try {
           val response = networkMightThrow(request, chain, cacheKey)
           if (response.statusCode in 200..299) {
             //  let HTTP errors through
             return response
           }
         } catch (e: ApolloException) {
+          // Original cause of network request failure
+          networkException = e
+        }
 
+        try {
+          return cacheMightThrow(request, cacheKey)
+        } catch (cacheMissException: HttpCacheMissException) {
+          // In case of exception thrown by network request,
+          // ApolloException is the root cause and cache miss exception will be suppressed
+          networkException?.addSuppressed(cacheMissException)
+
+          // Throw network exception, but in case of response status code not in 200..299 throw cache miss exception
+          throw networkException ?: cacheMissException
```

Comment:
Could we throw a `ApolloCompositeException` here instead (to be more consistent with what we do [here](https://github.com/apollographql/apollo-kotlin/blob/42e3d90abc1e8f75b794fc1440c61c1c14457d26/apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ClientCacheExtensions.kt#L180) and a few other places)?

Reply:
Here ApolloCompositeException only throws second
```` 
ApolloException(message = "multiple exceptions happened", second)
````
If we do this then clients must catch exception and then use first.addSuppressed(second) before re-throwing the exception otherwise there network failure will be masked by HttpCacheMissException

Reply:
What if we added the suppressed exception always? 

```kotlin
class ApolloCompositeException(first: Throwable?, second: Throwable?) : ApolloException(message = "multiple exceptions happened", second) {
  val first = (first as? ApolloException) ?: throw RuntimeException("unexpected first exception", first)
  val second = (second as? ApolloException) ?: throw RuntimeException("unexpected second exception", second)
  init {
    if (first != null) {
      addSuppressed(first)
    }
  }
}
``
`

TIL about suppressed exceptions but it looks like they're designed for exactly that case?

Reply:
Works

Reply:
Should i proceed with the above changes?

Reply:
We can't change the 'first' and 'second' property names nor types, that'd be a breaking change.

We could have just thrown the second one and add first as suppressed one but it's too late now because of API stability (and also because I'm not sure native and js have suppressed exceptions)

Id there a specific reason you're adding the suppressed exception to the second one (and not to the ApolloComposite one) ?

Reply:
ApolloCompositeException's cause is always 'second' that means if there is no 'second' exception both exceptions stacktrace will be lost

Reply:
For native and js this might help - https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/suppressed-exceptions.html

Reply:
> For native and js this might help

👍 Nice. Looks like it's working for JS and native too! Still not a change we can do today due to backward compatibility reasons

> ApolloCompositeException's cause is always 'second' that means if there is no 'second' exception both exceptions stacktrace will be lost

Isn't it possible to add suppressed exception on the exception itself (and not on its cause?)

Reply:
I have the impression that throwing `HttpCacheMissException` if there was no network exception, and `ApolloCompositeException` if there was one will allow the calling code to have all the information it needs? The code can check the type of the exception, and if it is composite, access `first` and `second`.  Am I missing something? Adding the suppressed in `ApolloCompositeException` is nice though!

Reply:
Yes it is possible
````
class ApolloCompositeException(
    original: Throwable?, suppressed: Throwable?,
) : ApolloException(message = "multiple exceptions happened", original ?: suppressed) {

  val first = (original as? ApolloException) ?: throw RuntimeException("unexpected first exception", original)
  val second = (suppressed as? ApolloException) ?: throw RuntimeException("unexpected second exception", suppressed)

  init {
    if (original != null && suppressed != null) addSuppressed(suppressed)
  }
}
````

Reply:
LGTM!